lib/Sema/CSSimplify.cpp

```diff
@@ -855,34 +848,23 @@ getCalleeDeclAndArgs(ConstraintSystem &cs,
   
   // If we didn't find any matching overloads, we're done.
   if (!choice)
-    return std::make_tuple(nullptr, /*hasCurriedSelf=*/false, argLabels,
+    return std::make_tuple(nullptr, /*hasAppliedSelf=*/false, argLabels,
                            hasTrailingClosure);
 
   // If there's a declaration, return it.
   if (auto *decl = choice->getDeclOrNull()) {
-    bool hasCurriedSelf = false;
-    if (decl->getDeclContext()->isTypeContext()) {
-      if (auto function = dyn_cast<AbstractFunctionDecl>(decl)) {
-        // References to instance members on a metatype stay at level 0.
-        // Everything else is level 1.
-        if (!(function->isInstanceMember() &&
-              cs.getFixedTypeRecursive(choice->getBaseType(),
-                                       /*wantRValue=*/true)
-                ->is<AnyMetatypeType>()))
-          hasCurriedSelf = true;
-      } else if (isa<SubscriptDecl>(decl)) {
-        // Subscript level 1 == the indices.
-        hasCurriedSelf = true;
-      } else if (isa<EnumElementDecl>(decl)) {
-        // Enum element level 1 == the payload.
-        hasCurriedSelf = true;
-      }
-    }
+    auto baseType = choice->getBaseType();
+    if (baseType)
+      baseType = cs.getFixedTypeRecursive(baseType, /*wantRValue=*/true);
 
-    return std::make_tuple(decl, hasCurriedSelf, argLabels, hasTrailingClosure);
+    // In most cases where we reference a declaration with a curried self
+    // parameter, it gets dropped from the type of the reference.
+    bool hasAppliedSelf =
+        decl->hasCurriedSelf() && doesDeclRefApplyCurriedSelf(baseType, decl);
+    return std::make_tuple(decl, hasAppliedSelf, argLabels, hasTrailingClosure);
```

Comment:
I wonder if it would be cleaner to return `choice` form here and add `baseType` parameter to `computeDefaultMap` so it could take care of figuring out whether self has been applied or not instead of relying on caller to do so?

Reply:
@xedin Thanks for the review! Unfortunately, I don't believe this will be easy for `coerceCallArguments`, which calls `computeDefaultMap` with slightly different logic for `skipCurriedSelf` – for example, passing `false` for a `SelfApplyExpr`.

Reply:
Fair enough.